rust/lance-io/src/scheduler.rs

```diff
@@ -3,42 +3,238 @@
 
+// Don't log backpressure warnings until at least this many seconds have passed
+const BACKPRESSURE_MIN: u64 = 5;
+// Don't log backpressure warnings more than once / minute
+const BACKPRESSURE_DEBOUNCE: u64 = 60;
+
+// We want to allow requests that have a lower priority than any
+// currently in-flight request.  This helps avoid potential deadlocks
+// related to backpressure.  Unfortunately, it is quite expensive to
+// keep track of which priorities are in-flight.
+//
+// TODO: At some point it would be nice if we can optimize this away but
+// in_flight should remain relatively small (generally less than 256 items)
+// and has not shown itself to be a bottleneck yet.
+struct PrioritiesInFlight {
+    in_flight: Vec<u128>,
+}
+
+impl PrioritiesInFlight {
+    fn new(capacity: u32) -> Self {
+        Self {
+            in_flight: Vec::with_capacity(capacity as usize * 2),
+        }
+    }
+
+    fn min_in_flight(&self) -> u128 {
+        self.in_flight.first().copied().unwrap_or(u128::MAX)
+    }
+
+    fn push(&mut self, prio: u128) {
+        let pos = match self.in_flight.binary_search(&prio) {
+            Ok(pos) => pos,
+            Err(pos) => pos,
+        };
+        self.in_flight.insert(pos, prio);
+    }
+
+    fn remove(&mut self, prio: u128) {
+        if let Ok(pos) = self.in_flight.binary_search(&prio) {
+            self.in_flight.remove(pos);
+        } else {
+            unreachable!();
+        }
+    }
+}
+
+struct IoQueueState {
+    // Number of IOPS we can issue concurrently before pausing I/O
+    iops_avail: u32,
+    // Number of bytes we are allowed to buffer in memory before pausing I/O
+    //
+    // This can dip below 0 due to I/O prioritization
+    bytes_avail: i64,
+    // Pending I/O requests
+    pending_requests: BinaryHeap<IoTask>,
+    // Priorities of in-flight requests
+    priorities_in_flight: PrioritiesInFlight,
+    // Set when the scheduler is finished to notify the I/O loop to shut down
+    closed: bool,
+    // Time when the scheduler started
+    start: Instant,
+    // Last time we warned about backpressure
+    last_warn: AtomicU64,
+}
+
+impl IoQueueState {
+    fn new(io_capacity: u32, io_buffer_size: u64) -> Self {
+        Self {
+            iops_avail: io_capacity,
+            bytes_avail: io_buffer_size as i64,
+            pending_requests: BinaryHeap::new(),
+            priorities_in_flight: PrioritiesInFlight::new(io_capacity),
+            closed: false,
+            start: Instant::now(),
+            last_warn: AtomicU64::from(0),
+        }
+    }
+
+    fn warn_if_needed(&self) {
+        let seconds_elapsed = self.start.elapsed().as_secs();
+        let last_warn = self.last_warn.load(Ordering::Acquire);
+        let since_last_warn = seconds_elapsed - last_warn;
+        if (last_warn == 0
+            && seconds_elapsed > BACKPRESSURE_MIN
+            && seconds_elapsed < BACKPRESSURE_DEBOUNCE)
+            || since_last_warn > BACKPRESSURE_DEBOUNCE
```

Comment:
I believe we always know `seconds_elapsed >= last_warn > BACKPRESSURE_MIN` if we are in the second branch so I think that is redundant.

Basically the first branch says "If we've never warned before then wait at least 5 seconds before we emit the first warning".  The second branch says "If we've waited at least a minute since the last warning (or a minute since the start if we've never warned) then go ahead and emit another warning"

Reply:
Oh I see, we don't hit the second branch unless we've warned at least once before, and thus have already passed the min time test. Nevermind.